File: apps/ios/Sources/Core/Navigation/PendingRoute.swift
import Foundation

// No URL scheme/associated-domain registration. Future adapters must inject a verified environment policy.
struct RoomRouteHint: Equatable, Sendable { let roomID: String }
struct ContentRouteParser: Sendable {
    let allowedHost: String
    let roomPathPrefix: String
    func parse(_ value: String) -> RoomRouteHint? {
        guard value.utf8.count <= 2048, let url = URLComponents(string: value),
              url.scheme == "https", url.host == allowedHost, url.port == nil,
              url.user == nil, url.password == nil, url.query == nil, url.fragment == nil,
              url.percentEncodedPath.hasPrefix(roomPathPrefix) else { return nil }
        let id = String(url.percentEncodedPath.dropFirst(roomPathPrefix.count))
        let allowed = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_-"
        guard (1...128).contains(id.utf8.count), id.allSatisfy({ allowed.contains($0) }) else { return nil }
        return RoomRouteHint(roomID: id)
    }
}
struct RouteTicket: Equatable, Sendable {
    let revision: UInt64
    let scope: UInt64
    let hint: RoomRouteHint
    fileprivate let eventID: String
    fileprivate let expiresAt: UInt64
}
// Value reducer; caller must keep it on one actor. It does not perform navigation or authorization.
struct PendingRouteQueue: Sendable {
    static let ttl: UInt64 = 300_000
    private var scope: UInt64 = 0
    private var revision: UInt64 = 0
    private var pending: RouteTicket?
    private var consumed: [(String, UInt64)] = []
    var scopeToken: UInt64 { scope }
    mutating func resetScope() { scope += 1; revision += 1; pending = nil; consumed.removeAll() }
    @discardableResult mutating func cancel(_ ticket: RouteTicket) -> Bool {
        guard pending == ticket, ticket.scope == scope else { return false }
        pending = nil; revision += 1
        return true
    }
    @discardableResult mutating func offer(_ hint: RoomRouteHint, eventID: String, now: UInt64, expectedScope: UInt64) -> Bool {
        guard expectedScope == scope else { return false }
        precondition(now <= UInt64.max - Self.ttl)
        guard !eventID.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty, eventID.count <= 256 else { return false }
        consumed.removeAll { $0.1 <= now }
        if consumed.contains(where: { $0.0 == eventID }) { return false }
        if let pending, pending.eventID == eventID, pending.expiresAt > now { return false }
        revision += 1
        pending = RouteTicket(revision: revision, scope: scope, hint: hint, eventID: eventID, expiresAt: now + Self.ttl)
        return true
    }
    mutating func begin(now: UInt64) -> RouteTicket? {
        if let current = pending, now >= current.expiresAt { pending = nil }
        return pending
    }
    mutating func consume(_ ticket: RouteTicket, now: UInt64) -> RoomRouteHint? {
        guard let current = begin(now: now), current == ticket, ticket.scope == scope else { return nil }
        pending = nil
        consumed.append((ticket.eventID, ticket.expiresAt))
        if consumed.count > 64 { consumed.removeFirst() }
        return ticket.hint
    }
}

File: apps/android/app/src/main/java/chat/rogi/rogichat/core/navigation/PendingRoute.kt
package chat.rogi.rogichat.core.navigation

import java.net.URI

// Not registered with Android intents. A verified environment policy must be injected by a future adapter.
data class RoomRouteHint(val roomId: String)
class ContentRouteParser(private val allowedHost: String, private val roomPathPrefix: String) {
    fun parse(value: String): RoomRouteHint? {
        if (value.length > 2048) return null
        val uri = try { URI(value) } catch (_: Exception) { return null }
        if (uri.scheme != "https" || uri.host != allowedHost || uri.port != -1 || uri.rawUserInfo != null ||
            uri.rawQuery != null || uri.rawFragment != null) return null
        val path = uri.rawPath ?: return null
        if (!path.startsWith(roomPathPrefix)) return null
        val id = path.removePrefix(roomPathPrefix)
        return if (id.matches(Regex("[A-Za-z0-9_-]{1,128}"))) RoomRouteHint(id) else null
    }
}

class RouteTicket internal constructor(val revision: Long, val scope: Long, val hint: RoomRouteHint,
                                            internal val eventId: String, internal val expiresAt: Long)

// Caller-confined reducer. Time must be monotonic; it never navigates or asserts server authorization.
class PendingRouteQueue {
    private var scope = 0L
    private var revision = 0L
    private var pending: RouteTicket? = null
    private val consumed = linkedMapOf<String, Long>()
    val scopeToken: Long get() = scope
    fun resetScope() { scope++; revision++; pending = null; consumed.clear() }
    fun cancel(ticket: RouteTicket): Boolean {
        if (pending != ticket || ticket.scope != scope) return false
        pending = null; revision++
        return true
    }
    fun offer(hint: RoomRouteHint, eventId: String, now: Long, expectedScope: Long): Boolean {
        if (expectedScope != scope) return false
        require(now >= 0 && now <= Long.MAX_VALUE - TTL)
        if (eventId.isBlank() || eventId.length > 256) return false
        consumed.entries.removeAll { it.value <= now }
        if (consumed.containsKey(eventId) || pending?.let { it.eventId == eventId && it.expiresAt > now } == true) return false
        revision++
        pending = RouteTicket(revision, scope, hint, eventId, now + TTL)
        return true
    }
    fun begin(now: Long): RouteTicket? {
        if (pending?.let { now >= it.expiresAt } == true) pending = null
        return pending
    }
    // Call after await: only the still-current ticket may be applied to the UI in the same turn.
    fun consume(ticket: RouteTicket, now: Long): RoomRouteHint? {
        val current = begin(now) ?: return null
        if (current != ticket || ticket.scope != scope) return null
        pending = null
        consumed[ticket.eventId] = ticket.expiresAt
        while (consumed.size > 64) consumed.remove(consumed.keys.first())
        return ticket.hint
    }
    companion object { const val TTL = 300_000L }
}
